File: src/main/java/com/example/document/SimpleDocumentationProvider.java
package com.example.document;

import com.intellij.lang.documentation.AbstractDocumentationProvider;
import com.intellij.lang.documentation.DocumentationMarkup;
import com.intellij.lang.javascript.psi.JSLiteralExpression;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.psi.PsiElement;
import com.intellij.psi.css.CssDeclaration;
import com.intellij.psi.css.CssRuleset;
import com.intellij.psi.css.CssSelector;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class SimpleDocumentationProvider extends AbstractDocumentationProvider {

    public static void renderCssBlock(StringBuilder sb, CssDeclaration[] cssDeclarations) {
        for (CssDeclaration declaration : cssDeclarations) {
            if (declaration != null) {
                sb.append(DocumentationMarkup.PRE_ELEMENT.addText(StringUtil.trim(declaration.getText())));
            }
        }
    }

    @Override
    public @Nullable String generateDoc(PsiElement element, @Nullable PsiElement originalElement) {
        StringBuilder content = new StringBuilder();
        content.append(DocumentationMarkup.DEFINITION_START);
        if (originalElement != null && originalElement.getParent() instanceof JSLiteralExpression js) {
            content.append(".").append(js.getStringValue());
            content.append(DocumentationMarkup.DEFINITION_END);
            content.append("<hr/>");
            content.append(DocumentationMarkup.SECTIONS_START);
            if (js.getReference() != null && js.getReference().resolve() instanceof CssSelector cssSelector) {
                CssRuleset ruleset = cssSelector.getRuleset();
                if (ruleset != null && ruleset.getBlock() != null) {
                    renderCssBlock(content, ruleset.getBlock().getDeclarations());
                }
            }
            content.append(DocumentationMarkup.SECTION_END);
        }
        return String.valueOf(content);
    }

    @Override
    public @Nullable @Nls String generateHoverDoc(@NotNull PsiElement element, @Nullable PsiElement originalElement) {
        return generateDoc(element, originalElement);
    }

}

File: src/main/kotlin/com/peppa/css/document/SimpleDocumentationProvider.kt
package com.peppa.css.document

import com.intellij.lang.css.CSSLanguage
import com.intellij.lang.documentation.AbstractDocumentationProvider
import com.intellij.lang.documentation.QuickDocHighlightingHelper.appendStyledCodeBlock
import com.intellij.lang.javascript.psi.JSLiteralExpression
import com.intellij.lang.javascript.psi.JSReferenceExpression
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.css.CssRuleset

class SimpleDocumentationProvider : AbstractDocumentationProvider() {

    override fun getCustomDocumentationElement(
        editor: Editor,
        file: PsiFile,
        context: PsiElement?,
        targetOffset: Int
    ): PsiElement? {
        if (context == null) return null
        val parent = context.parent ?: context
        for (ref in parent.references) {
            val resolved = ref.resolve()
            if (resolved is CssRuleset) return resolved
        }
        for (ref in context.references) {
            val resolved = ref.resolve()
            if (resolved is CssRuleset) return resolved
        }
        return null
    }

    override fun getQuickNavigateInfo(element: PsiElement?, originalElement: PsiElement?): String? {
        if (element !is CssRuleset || !isValidContext(originalElement)) return null
        return renderDoc(element)
    }

    private fun isValidContext(context: PsiElement?): Boolean =
        context is JSLiteralExpression
                || context?.parent is JSLiteralExpression
                || context is JSReferenceExpression
                || context?.parent is JSReferenceExpression

    private fun renderDoc(cssRuleset: CssRuleset): String {
        val text = cssRuleset.text.trimIndent()
        return StringBuilder()
            .appendStyledCodeBlock(cssRuleset.project, CSSLanguage.INSTANCE, code = text)
            .toString()
    }
}
